// @darraghor/eslint-plugin-nestjs-typed/dist/rules/apiMethodsShouldSpecifyApiOperation/apiMethodsShouldSpecifyApiOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shouldUseApiResponseDecorator = void 0;
const createRule_1 = require("../../utils/createRule");
const typedTokenHelpers_1 = require("../../utils/typedTokenHelpers");
const shouldUseApiResponseDecorator = (node) => {
    const hasApiMethodDecorator = typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(node, ["Get", "Post", "Put", "Delete", "Patch", "Options", "Head", "All"]);
    const hasApiOperationDecorator = typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(node, ["ApiOperation"]);
    return hasApiMethodDecorator && !hasApiOperationDecorator;
};
exports.shouldUseApiResponseDecorator = shouldUseApiResponseDecorator;
const rule = (0, createRule_1.createRule)({
    name: "api-method-should-specify-api-operation",
    meta: {
        docs: {
            description: "Api methods should at least specify the expected ApiOperation.",
            requiresTypeChecking: false,
        },
        messages: {
            shouldSpecifyApiOperation: `A method decorated with @Get, @Post etc. should specify the expected ApiOperation e.g. @ApiOperation({description: ""}). These decorators are in the @nestjs/swagger npm package.`,
        },
        schema: [],
        hasSuggestions: false,
        type: "suggestion",
    },
    defaultOptions: [],
    create(context) {
        return {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            MethodDefinition(node) {
                if ((0, exports.shouldUseApiResponseDecorator)(node)) {
                    context.report({
                        node: node,
                        messageId: "shouldSpecifyApiOperation",
                    });
                }
            },
        };
    },
});
exports.default = rule;
