// @darraghor/eslint-plugin-nestjs-typed/dist/utils/nestModules/nestProvidedInjectableMapper.testData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.moduleMappingTestData = exports.fakeContext = exports.fakeFilePath = void 0;
const NestProvidedInjectablesMap_1 = require("./models/NestProvidedInjectablesMap");
exports.fakeFilePath = "fake/path.ts";
// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
exports.fakeContext = {
    parserOptions: {
        ecmaVersion: 2019,
        ecmaFeatures: { globalReturn: false },
        sourceType: "module",
    },
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
};
exports.moduleMappingTestData = [
    {
        moduleCode: `import { Module } from "@nestjs/common";

    @Module({
        controllers: [MyController],
        providers: [
            MyProvider,
            MyInjectable,
        ],
        imports: [MyExternalModule, MySecondModule],
        exports: [MyInjectable],
    })
    export class MyModule {}`,
        expectedMapping: [
            exports.fakeFilePath,
            new NestProvidedInjectablesMap_1.NestProvidedInjectablesMap(new Set(["MyController"]), new Set(["MyProvider", "MyInjectable"])),
        ],
    },
    {
        moduleCode: `import { Provider } from "@nestjs/common";
import { MyOtherInjectable } from "./config/MyOtherInjectable";

export const MyOtherInjectableProvider: Provider = {
    provide: MyOtherInjectable,
    useFactory: async (
        config: MyService
    ): Promise<MyOtherInjectable> => {
        return new MyOtherInjectable()
    },
    inject: [MyService],
};`,
        expectedMapping: [
            exports.fakeFilePath,
            new NestProvidedInjectablesMap_1.NestProvidedInjectablesMap(new Set([]), new Set(["MyOtherInjectable"])),
        ],
    },
];
