// @darraghor/eslint-plugin-nestjs-typed/dist/rules/noDuplicateDecorators/noDuplicateDecorators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.standardDecoratorsToValidate = exports.shouldTrigger = void 0;
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable unicorn/prevent-abbreviations */
const createRule_1 = require("../../utils/createRule");
const utils_1 = require("@typescript-eslint/utils");
const shouldTrigger = () => {
    return true;
};
exports.shouldTrigger = shouldTrigger;
exports.standardDecoratorsToValidate = ["Controller", "Injectable"];
const rule = (0, createRule_1.createRule)({
    name: "no-duplicate-decorators",
    meta: {
        docs: {
            description: "Some decorators should only be used once on a property or class. This rule enforces that.",
            requiresTypeChecking: false,
        },
        messages: {
            noDuplicateDecorators: "You have listed the same decorator more than once. Was this intentional?",
        },
        schema: [
            {
                type: "object",
                properties: {
                    customList: {
                        description: "A list of custom decorators that this rule will validate for duplicates",
                        type: "array",
                        minItems: 0,
                        items: {
                            type: "string",
                            minLength: 1,
                        },
                    },
                },
            },
        ],
        hasSuggestions: true,
        type: "suggestion",
    },
    defaultOptions: [
        {
            customList: new Array(),
        },
    ],
    create(context) {
        const customListArrayItem = context.options?.[0];
        let decoratorsToValidate = customListArrayItem?.customList;
        if (!customListArrayItem?.customList ||
            customListArrayItem?.customList.length === 0) {
            decoratorsToValidate = exports.standardDecoratorsToValidate;
        }
        return {
            ["PropertyDefinition,ClassDeclaration"](node) {
                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
                const allDecorators = node?.decorators;
                if (allDecorators && allDecorators.length > 1) {
                    const decoratorNames = allDecorators.map((decorator) => {
                        if (decorator.expression.type ===
                            utils_1.TSESTree.AST_NODE_TYPES.CallExpression &&
                            decorator.expression.callee &&
                            decorator.expression.callee.type ===
                                utils_1.TSESTree.AST_NODE_TYPES.Identifier) {
                            // eslint-disable-next-line @typescript-eslint/no-unsafe-return
                            return decorator.expression.callee.name;
                        }
                        return "";
                    });
                    const duplicateDecorators = decoratorNames.filter((decoratorName, index) => {
                        return (decoratorName &&
                            decoratorNames.indexOf(decoratorName) !== index);
                    });
                    if (duplicateDecorators.length > 0 &&
                        decoratorsToValidate.some((decoratorToValidate) => duplicateDecorators.includes(decoratorToValidate))) {
                        context.report({
                            node: node,
                            messageId: "noDuplicateDecorators",
                            suggest: [
                                {
                                    messageId: "noDuplicateDecorators",
                                    fix: (fixer) => {
                                        const decoratorsToKeep = decoratorNames.filter((decoratorName, index) => {
                                            return (decoratorName &&
                                                decoratorNames.indexOf(decoratorName) === index);
                                        });
                                        const decoratorsToKeepText = decoratorsToKeep.join(", ");
                                        return fixer.replaceText(node, decoratorsToKeepText);
                                    },
                                },
                            ],
                        });
                    }
                }
            },
        };
    },
});
exports.default = rule;
