// @darraghor/eslint-plugin-nestjs-typed/dist/rules/apiMethodsShouldSpecifyApiResponse/apiMethodsShouldSpecifyApiResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shouldUseApiResponseDecorator = void 0;
const createRule_1 = require("../../utils/createRule");
const typedTokenHelpers_1 = require("../../utils/typedTokenHelpers");
const shouldUseApiResponseDecorator = (node) => {
    const hasApiMethodDecorator = typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(node, ["Get", "Post", "Put", "Delete", "Patch", "Options", "Head", "All"]);
    const hasApiResponseDecorator = typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(node, [
        "ApiResponse",
        "ApiOkResponse",
        "ApiCreatedResponse",
        "ApiAcceptedResponse",
        "ApiNoContentResponse",
        "ApiMovedPermanentlyResponse",
        "ApiFoundResponse",
        "ApiBadRequestResponse",
        "ApiUnauthorizedResponse",
        "ApiTooManyRequestsResponse",
        "ApiNotFoundResponse",
        "ApiInternalServerErrorResponse",
        "ApiBadGatewayResponse",
        "ApiConflictResponse",
        "ApiForbiddenResponse",
        "ApiGatewayTimeoutResponse",
        "ApiGoneResponse",
        "ApiMethodNotAllowedResponse",
        "ApiNotAcceptableResponse",
        "ApiNotImplementedResponse",
        "ApiPreconditionFailedResponse",
        "ApiPayloadTooLargeResponse",
        "ApiRequestTimeoutResponse",
        "ApiServiceUnavailableResponse",
        "ApiUnprocessableEntityResponse",
        "ApiUnsupportedMediaTypeResponse",
        "ApiDefaultResponse",
    ]);
    return hasApiMethodDecorator && !hasApiResponseDecorator;
};
exports.shouldUseApiResponseDecorator = shouldUseApiResponseDecorator;
const rule = (0, createRule_1.createRule)({
    name: "api-method-should-specify-api-response",
    meta: {
        docs: {
            description: "Api methods should at least specify the expected OK response with @ApiOkResponse. But also add any error responses that might not be expected (e.g. 429)",
            requiresTypeChecking: false,
        },
        messages: {
            shouldSpecifyApiResponse: `A method decorated with @Get, @Post etc. should specify the expected ApiResponse e.g. @ApiOkResponse(type: MyType). These decorators are in the @nestjs/swagger npm package.`,
        },
        schema: [],
        hasSuggestions: false,
        type: "suggestion",
    },
    defaultOptions: [],
    create(context) {
        return {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            MethodDefinition(node) {
                if ((0, exports.shouldUseApiResponseDecorator)(node)) {
                    context.report({
                        node: node,
                        messageId: "shouldSpecifyApiResponse",
                    });
                }
            },
        };
    },
});
exports.default = rule;
