// @darraghor/eslint-plugin-nestjs-typed/dist/rules/paramDecoratorNameMatchesRouteParam/paramDecoratorNameMatchesRouteParam.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shouldTrigger = exports.isParameterNameIncludedInAPathPart = exports.hasPathPartsAnyRegexParams = exports.parsePathParts = void 0;
/* eslint-disable unicorn/prevent-abbreviations */
const utils_1 = require("@typescript-eslint/utils");
const createRule_1 = require("../../utils/createRule");
const nestRequestMethodDecoratorNames = new Set([
    "Get",
    "Post",
    "Put",
    "Delete",
    "Patch",
    "Options",
    "Head",
    "All",
]);
const parsePathParts = (decorator) => {
    const decoratorArgument = decorator?.expression
        ?.arguments[0];
    if (decoratorArgument?.type === utils_1.TSESTree.AST_NODE_TYPES.TemplateLiteral ||
        decoratorArgument?.type === utils_1.TSESTree.AST_NODE_TYPES.Identifier) {
        return ["dareslint__skip"];
    }
    if (decoratorArgument?.type === utils_1.TSESTree.AST_NODE_TYPES.Literal) {
        return [decoratorArgument.raw];
    }
    if (decoratorArgument?.type === utils_1.TSESTree.AST_NODE_TYPES.ArrayExpression) {
        return decoratorArgument.elements.map((x) => x.raw);
    }
    if (decoratorArgument?.type === utils_1.TSESTree.AST_NODE_TYPES.ObjectExpression) {
        return decoratorArgument.properties
            .filter((x) => x?.key
            ?.name === "path")
            .map((x) => x.value.raw);
    }
    return [];
};
exports.parsePathParts = parsePathParts;
/**
 * nestjs allows for paths with _+?()*
 * this rule doesn't support parsing those so we'll just pass
 */
const hasPathPartsAnyRegexParams = (pathPartsToCheck) => {
    // prettier-ignore
    // eslint-disable-next-line no-useless-escape
    const specialCharacterRegex = /(dareslint__skip|\*|\+|\?|\(|\)|_)/; //new RegExp("([\?\+\*\_\(\)])")
    return pathPartsToCheck.some((pathPart) => {
        return specialCharacterRegex.test(pathPart);
    });
};
exports.hasPathPartsAnyRegexParams = hasPathPartsAnyRegexParams;
/**
 * Checks if there is a matching path part for the paramName
 * @param paramName
 * @param pathPartsToCheck
 * @returns
 */
const isParameterNameIncludedInAPathPart = (paramName, pathPartsToCheck) => {
    return pathPartsToCheck.some((pathPart) => {
        return (
        // note to reader: this might be better as a regex. feel free to open a pr!
        pathPart === `":${paramName}"` ||
            pathPart === `':${paramName}'` ||
            pathPart.includes(`/:${paramName}/`) ||
            pathPart.includes(`/:${paramName}"`) ||
            pathPart.includes(`":${paramName}/`) ||
            pathPart.includes(`/:${paramName}'`) ||
            pathPart.includes(`':${paramName}/`));
    });
};
exports.isParameterNameIncludedInAPathPart = isParameterNameIncludedInAPathPart;
const shouldTrigger = (decorator) => {
    if (!decorator) {
        return {
            hasColonInName: false,
            paramNameNotMatchedInPath: false,
        };
    }
    // grab the param name
    const paramName = decorator.expression
        ?.arguments[0]?.value;
    // if there's no param name get out of here
    if (!paramName || paramName === "") {
        return {
            hasColonInName: false,
            paramNameNotMatchedInPath: false,
        };
    }
    // param names don't need the colon
    if (paramName.startsWith(":")) {
        return {
            hasColonInName: true,
            paramNameNotMatchedInPath: false,
        };
    }
    let pathPartsToCheck = [];
    // grab any controller path parts
    // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
    const controllerDecorator = decorator.parent?.parent?.parent?.parent
        ?.parent?.decorators?.find((d) => {
        return (d.expression
            .callee?.name === "Controller");
    });
    pathPartsToCheck = pathPartsToCheck.concat((0, exports.parsePathParts)(controllerDecorator));
    // grab any api method path parts from method decorator
    const methodDefinition = decorator.parent?.parent
        ?.parent;
    // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
    const methodDecorator = methodDefinition?.decorators?.find((d) => {
        return nestRequestMethodDecoratorNames.has(d.expression
            .callee?.name);
    });
    pathPartsToCheck = pathPartsToCheck.concat((0, exports.parsePathParts)(methodDecorator));
    const shouldIgnoreThisSetOfRoutes = 
    // is a template literal argument
    // is an identifier argument
    (0, exports.hasPathPartsAnyRegexParams)(pathPartsToCheck);
    if (shouldIgnoreThisSetOfRoutes) {
        return {
            hasColonInName: false,
            paramNameNotMatchedInPath: false,
        };
    }
    // check that the param name is in one path part
    return {
        hasColonInName: false,
        paramNameNotMatchedInPath: !(0, exports.isParameterNameIncludedInAPathPart)(paramName, pathPartsToCheck),
    };
};
exports.shouldTrigger = shouldTrigger;
const rule = (0, createRule_1.createRule)({
    name: "param-decorator-name-matches-route-param",
    meta: {
        docs: {
            description: 'Param decorators with a name parameter e.g. Param("myvar") should match a specified route parameter - e.g. Get(":myvar")',
            requiresTypeChecking: false,
        },
        messages: {
            paramIdentifierDoesntNeedColon: "You don't need to specify the colon (:) in a Param decorator",
            paramIdentifierShouldMatch: 'Param decorators with identifiers e.g. Param("myvar") should match a specified route parameter - e.g. Get(":myvar")',
        },
        schema: [],
        hasSuggestions: false,
        type: "suggestion",
    },
    defaultOptions: [],
    create(context) {
        return {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            Decorator(node) {
                if (node.expression
                    ?.callee?.name !== "Param") {
                    return;
                }
                // eslint-disable-next-line @typescript-eslint/naming-convention
                const result = (0, exports.shouldTrigger)(node);
                if (result.paramNameNotMatchedInPath) {
                    context.report({
                        node: node,
                        messageId: "paramIdentifierShouldMatch",
                    });
                }
                if (result.hasColonInName) {
                    context.report({
                        node: node,
                        messageId: "paramIdentifierDoesntNeedColon",
                    });
                }
            },
        };
    },
});
exports.default = rule;
