// @darraghor/eslint-plugin-nestjs-typed/dist/rules/apiMethodsShouldBeGuarded/apiMethodsShouldBeGuarded.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiMethodsShouldBeGuarded = void 0;
const utils_1 = require("@typescript-eslint/utils");
const createRule_1 = require("../../utils/createRule");
const typedTokenHelpers_1 = require("../../utils/typedTokenHelpers");
const apiMethodsShouldBeGuarded = (node) => {
    const hasApiMethodDecorator = typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(node, ["Get", "Post", "Put", "Delete", "Patch", "Options", "Head", "All"]);
    const hasUseGuardsDecorator = typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(node, ["UseGuards"]);
    function findClassDeclaration(node) {
        if (node.type === utils_1.TSESTree.AST_NODE_TYPES.ClassDeclaration) {
            return node;
        }
        if (node.parent) {
            return findClassDeclaration(node.parent);
        }
        return null;
    }
    const classNode = findClassDeclaration(node);
    const hasUseGuardsDecoratorOnController = classNode
        ? typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(classNode, ["UseGuards"])
        : false;
    return (hasApiMethodDecorator &&
        !hasUseGuardsDecorator &&
        !hasUseGuardsDecoratorOnController);
};
exports.apiMethodsShouldBeGuarded = apiMethodsShouldBeGuarded;
const rule = (0, createRule_1.createRule)({
    name: "api-methods-should-be-guarded",
    meta: {
        docs: {
            description: "Endpoints should have authentication guards to maintain our security model.",
            requiresTypeChecking: false,
        },
        messages: {
            apiMethodsShouldBeGuarded: "All controller endpoints should have @UseGuards decorators, or one decorating the root of the Controller.",
        },
        schema: [],
        hasSuggestions: false,
        type: "suggestion",
    },
    defaultOptions: [],
    create(context) {
        return {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            MethodDefinition(node) {
                if ((0, exports.apiMethodsShouldBeGuarded)(node)) {
                    context.report({
                        node: node,
                        messageId: "apiMethodsShouldBeGuarded",
                    });
                }
            },
        };
    },
});
exports.default = rule;
