// @darraghor/eslint-plugin-nestjs-typed/dist/rules/controllerDecoratedHasApiTags/controllerDecoratedHasApiTags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shouldUseApiTagDecorator = void 0;
const createRule_1 = require("../../utils/createRule");
const typedTokenHelpers_1 = require("../../utils/typedTokenHelpers");
const shouldUseApiTagDecorator = (node) => {
    const hasControllerDecorator = typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(node, ["Controller"]);
    const hasApiTagDecorator = typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(node, [
        "ApiTags",
    ]);
    return hasControllerDecorator && !hasApiTagDecorator;
};
exports.shouldUseApiTagDecorator = shouldUseApiTagDecorator;
const rule = (0, createRule_1.createRule)({
    name: "controllers-should-supply-api-tags",
    meta: {
        docs: {
            description: "Controllers should supply an ApiTag to make swagger UI easier to navigate",
            requiresTypeChecking: false,
        },
        messages: {
            shouldUseApiTagDecorator: `Controllers should use @ApiTags decorator. This makes it much easier to navigate swagger UI. This ApiTags decorator is in the @nestjs/swagger package on npm.`,
        },
        schema: [],
        hasSuggestions: false,
        type: "suggestion",
    },
    defaultOptions: [],
    create(context) {
        return {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            ClassDeclaration(node) {
                if ((0, exports.shouldUseApiTagDecorator)(node)) {
                    context.report({
                        node: node,
                        messageId: "shouldUseApiTagDecorator",
                    });
                }
            },
        };
    },
});
exports.default = rule;
