// @darraghor/eslint-plugin-nestjs-typed/dist/rules/apiEnumPropertyBestPractices/apiEnumPropertyBestPractices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.needsEnumNameMatchingEnumType = exports.hasEnumSpecifiedCorrectly = void 0;
const utils_1 = require("@typescript-eslint/utils");
const createRule_1 = require("../../utils/createRule");
const typedTokenHelpers_1 = require("../../utils/typedTokenHelpers");
const enumTestResultModel_1 = require("./enumTestResultModel");
const noChangesRequiredResult = new enumTestResultModel_1.EnumTestResultModel({
    needsEnumAdded: false,
    needsEnumNameAdded: false,
    needsEnumNameToMatchEnumType: false,
    needsTypeRemoved: false,
});
const hasEnumSpecifiedCorrectly = (node, isEnumType) => {
    // is this an enum
    if (!isEnumType) {
        return noChangesRequiredResult;
    }
    // is this decorated with api documentation
    const decorators = typedTokenHelpers_1.typedTokenHelpers.getDecoratorsNamed(node, ["ApiPropertyOptional", "ApiProperty"]);
    if (decorators.length === 0) {
        return noChangesRequiredResult;
    }
    // check if there is an enum property in the provided options (enums should specify the enum property)
    const firstArgument = decorators[0].expression
        .arguments[0];
    if (!firstArgument) {
        return new enumTestResultModel_1.EnumTestResultModel({
            needsEnumAdded: true,
            needsEnumNameAdded: true,
            needsEnumNameToMatchEnumType: false,
            needsTypeRemoved: false,
        });
    }
    const enumProperty = firstArgument.properties.find((p) => p.key.name ===
        "enum");
    // check if there is a type: property in the provided options (enums shouldn't specify type)
    const hasTypeProperty = firstArgument.properties.find((p) => p.key.name ===
        "type") !== undefined;
    // check if there is an enumName: property in the provided options (enums should specify a name)
    const enumNameProperty = firstArgument.properties.find((p) => p.key.name ===
        "enumName");
    return new enumTestResultModel_1.EnumTestResultModel({
        needsEnumAdded: enumProperty === undefined,
        needsEnumNameAdded: enumNameProperty === undefined,
        needsEnumNameToMatchEnumType: (0, exports.needsEnumNameMatchingEnumType)(enumNameProperty, enumProperty),
        needsTypeRemoved: hasTypeProperty,
    });
};
exports.hasEnumSpecifiedCorrectly = hasEnumSpecifiedCorrectly;
const needsEnumNameMatchingEnumType = (enumNameProperty, enumProperty) => {
    // if enum props aren't specified we don't care about this scenario
    if (enumNameProperty === undefined || enumProperty === undefined) {
        return false;
    }
    const isEnumNameMatchingEnumType = enumNameProperty.value.value ===
        enumProperty.value.name;
    return !isEnumNameMatchingEnumType;
};
exports.needsEnumNameMatchingEnumType = needsEnumNameMatchingEnumType;
const rule = (0, createRule_1.createRule)({
    name: "api-enum-property-best-practices",
    meta: {
        docs: {
            description: "Enums should use the best practices for api documentation",
            requiresTypeChecking: false,
        },
        messages: {
            needsEnumNameAdded: `Properties with enum should also specify an enumName property to keep generated models clean`,
            needsTypeRemoved: `Properties with enum should not specify a type property`,
            enumNameShouldMatchType: `The enumName should match the enum type provided`,
        },
        schema: [],
        hasSuggestions: false,
        type: "suggestion",
    },
    defaultOptions: [],
    create(context) {
        //const globalScope = context.getScope();
        const parserServices = utils_1.ESLintUtils.getParserServices(context);
        const typeChecker = parserServices.program.getTypeChecker();
        return {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            PropertyDefinition(node) {
                const mappedNode = parserServices.esTreeNodeToTSNodeMap.get(node);
                const objectType = typeChecker.getTypeAtLocation(mappedNode);
                const isEnumType = typedTokenHelpers_1.typedTokenHelpers.isEnumType(objectType);
                const result = (0, exports.hasEnumSpecifiedCorrectly)(node, isEnumType);
                if (result.needsEnumNameAdded) {
                    context.report({
                        node: node,
                        messageId: "needsEnumNameAdded",
                    });
                }
                if (result.needsTypeRemoved) {
                    context.report({
                        node: node,
                        messageId: "needsTypeRemoved",
                    });
                }
                if (result.needsEnumNameToMatchEnumType) {
                    context.report({
                        node: node,
                        messageId: "enumNameShouldMatchType",
                    });
                }
            },
        };
    },
});
exports.default = rule;
